### PR TITLE
route: dedicated net ns for each subtest of runListRules

### LINKED ISSUE #23869: CI: route_linux_test.go:234: RouteSuitePrivileged.TestListRules
### Stacktrace

```
	 [10:45:22] START: route_linux_test.go:234: RouteSuitePrivileged.TestListRules
	 [10:45:22] Inside new network ns
	 [10:45:22] route_linux_test.go:235:
	 [10:45:22]     testListRules4(c)
	 [10:45:22] route_linux_test.go:487:
	 [10:45:22]     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 [10:45:22] ... Error: expected len: 3, got: 5
	 [10:45:22]   []netlink.Rule{
	 [10:45:22]   	{
	 [10:45:22] - 		Priority: -1,
	 [10:45:22] + 		Priority: 9,
	 [10:45:22]   		Family:   0,
	 [10:45:22] - 		Table:    255,
	 [10:45:22] + 		Table:    2004,
	 [10:45:22] - 		Mark:     -1,
	 [10:45:22] + 		Mark:     512,
	 [10:45:22] - 		Mask:     -1,
	 [10:45:22] + 		Mask:     3840,
	 [10:45:22]   		Tos:      0,
	 [10:45:22]   		TunID:    0,
	 [10:45:22]   		... // 12 identical fields
	 [10:45:22]   	},
	 [10:45:22] + 	s"ip rule 10: from all to all table 2005",
	 [10:45:22] + 	s"ip rule 100: from all to all table 255",
	 [10:45:22]   	{Priority: 32766, Table: 254, Mark: -1, Mask: -1, ...},
	 [10:45:22]   	{Priority: 32767, Table: 253, Mark: -1, Mask: -1, ...},
	 [10:45:22]   }
	 [10:45:22] 
	 [10:45:22] 
	 [10:45:22] 
	 [10:45:22] Closed new network ns
	 [10:45:22] Set back to previous network ns
	 [10:45:22] Inside new network ns
	 [10:45:22] Closed new network ns
	 [10:45:22] Set back to previous network ns
	 [10:45:22] FAIL: route_linux_test.go:234: RouteSuitePrivileged.TestListRules
```

Jenkins: https://jenkins.cilium.io/job/Cilium-PR-Runtime-net-next/5091/testReport/junit/(root)/Suite-runtime/RuntimePrivilegedUnitTests_Run_Tests/

Sysdump: 
[42153405_RuntimePrivilegedUnitTests_Run_Tests.zip](https://github.com/cilium/cilium/files/10776202/42153405_RuntimePrivilegedUnitTests_Run_Tests.zip)

### PR DESCRIPTION
**Pull request by mhofstetter**
Currently, there are cases where the test TestListRules/return_all_rules fails with the following error:

```
--- FAIL: TestListRules (0.02s)
    --- FAIL: TestListRules/returns_all_rules#01 (0.00s)
        route_linux_test.go:490: expected len: 2, got: 3
              []netlink.Rule{
              {
            - Priority: -1,
            + Priority: 9,
              Family:   10,
            - Table:    255,
            + Table:    2004,
            - Mark:     -1,
            + Mark:     512,
            - Mask:     -1,
            + Mask:     3840,
              Tos:      0,
              TunID:    0,
              ... // 11 identical fields
              IPProto:  0,
              UIDRange: nil,
            - Protocol: 2,
            + Protocol: 0,
              },
            + s"ip rule 100: from all to all table 255",
              {Priority: 32766, Family: 10, Table: 254, Mark: -1, ...},
              }
```

It looks like there's a switch of the network namespace during the test execution.

Therefore, this commit locks the OS thread for the execution of the test that runs in a dedicated network namespace. In addition, each sub-test of the table driven testset runs in its own network namespace because they run in their own go-routine.

https://pkg.go.dev/github.com/vishvananda/netns#pkg-overview

> ... Note that the current namespace is thread local so actions that set and reset namespaces should use LockOSThread to make sure the namespace doesn't change due to a goroutine switch. ...

Fixes: #23869

### CODE REVIEW

**Comment by mhofstetter** [comment]
Rebased to `main` to get fix for persistent issue in check `ci-ipsec-upgrade`: https://github.com/cilium/cilium/issues/29666

File: pkg/datapath/linux/route/route_linux_test.go [hunk 1]
```diff
@@ -275,6 +262,7 @@ func runListRules(t *testing.T, family int, fakeIP, fakeIP2 *net.IPNet) {
 			preRun:     func() *netlink.Rule { return nil },
 			postRun:    func(r *netlink.Rule) {},
 			setupWant: func(_ *netlink.Rule) ([]netlink.Rule, bool) {
+				defaultRules, _ := ListRules(family, nil)
```
Comment (gentoo-root) [inline pkg/datapath/linux/route/route_linux_test.go:265]: Off-topic, but it looks to me that this test isn't testing anything, it just compares `ListRules(family, nil)` with `ListRules(family, nil)`.

**Review by gentoo-root — APPROVED** [review]
Looks good.

File: pkg/datapath/linux/route/route_linux_test.go (hunk 1 above)
Comment (mhofstetter) [reply pkg/datapath/linux/route/route_linux_test.go:265]: agree. maybe the original intention was also the documentation aspect (listing without filtering) :shrug: 

i simply focused on fixing the flake.

**Pull request merged**